core/src/main/scala/kafka/tools/GetOffsetShell.scala

```diff
@@ -132,23 +161,85 @@ object GetOffsetShell {
+      case (tp, offset) => println(s"${tp.topic}:${tp.partition}:${Option(offset).getOrElse("")}")
     }
+  }
 
+  def compareTopicPartitions(a: TopicPartition, b: TopicPartition): Boolean = {
+    (a.topic(), a.partition()) < (b.topic(), b.partition())
   }
 
   /**
-   * Return the partition infos for `topic`. If the topic does not exist, `None` is returned.
+   * Creates a topic-partition filter based on a list of patterns.
+   * Expected format:
+   * List: TopicPartitionPattern(, TopicPartitionPattern)*
+   * TopicPartitionPattern: TopicPattern(:PartitionPattern)? | :PartitionPattern
+   * TopicPattern: REGEX
+   * PartitionPattern: NUMBER | NUMBER-(NUMBER)? | -NUMBER
    */
-  private def listPartitionInfos(consumer: KafkaConsumer[_, _], topic: String, partitionIds: Set[Int]): Option[Seq[PartitionInfo]] = {
-    val partitionInfos = consumer.listTopics.asScala.filter { case (k, _) => k == topic }.values.flatMap(_.asScala).toBuffer
-    if (partitionInfos.isEmpty)
-      None
-    else if (partitionIds.isEmpty)
-      Some(partitionInfos)
-    else
-      Some(partitionInfos.filter(p => partitionIds.contains(p.partition)))
+  def createTopicPartitionFilterWithPatternList(topicPartitions: String, excludeInternalTopics: Boolean): PartitionInfo => Boolean = {
+    val ruleSpecs = topicPartitions.split(",")
+    val rules = ruleSpecs.map { ruleSpec => parseRuleSpec(ruleSpec, excludeInternalTopics) }
+    tp => rules.exists(rule => rule.apply(tp))
   }
 
+  def parseRuleSpec(ruleSpec: String, excludeInternalTopics: Boolean): PartitionInfo => Boolean = {
+    def wrapNullOrEmpty(s: String): Option[String] = {
+      if (s == null || s.isEmpty)
+        None
+      else
+        Some(s)
+    }
+
+    val matcher = topicPartitionPattern.matcher(ruleSpec)
+    if (!matcher.matches() || matcher.groupCount() == 0)
+      throw new IllegalArgumentException(s"Invalid rule specification: $ruleSpec")
+
+    val topicPattern = wrapNullOrEmpty(matcher.group(1))
+    val lowerRange = if (matcher.groupCount() >= 2) wrapNullOrEmpty(matcher.group(2)) else None
+    val upperRangeSection = if (matcher.groupCount() >= 3) matcher.group(3) else null
+    val upperRange = wrapNullOrEmpty(upperRangeSection)
+    val isRange = upperRangeSection != null
+
+    val includeList = IncludeList(topicPattern.getOrElse(".*"))
+
+    val partitionFilter: Int => Boolean = if (lowerRange.isEmpty && upperRange.isEmpty) {
+      _ => true
+    } else if (lowerRange.isEmpty) {
+      val upperBound = upperRange.get.toInt
+      p => p < upperBound
+    } else if (upperRange.isEmpty) {
+      val lowerBound = lowerRange.get.toInt
+      if (isRange) {
+        p => p >= lowerBound
+      }
+      else {
+        p => p == lowerBound
+      }
+    } else {
+      val upperBound = upperRange.get.toInt
+      val lowerBound = lowerRange.get.toInt
+      p => p >= lowerBound && p < upperBound
+    }
```

Comment:
I think that we could further simplify this by differentiating the range case from the single partition case in the regex directly. We could for instance use the following regex: `"([^:,]*)(?::(?:([0-9]*)|(?:([0-9]*)-([0-9]*))))?"`.

It allows to do something like this for the processing:
```
    val matcher = topicPartitionPattern.matcher(ruleSpec)
    if (!matcher.matches())
      throw new IllegalArgumentException(s"Invalid rule specification: $ruleSpec")

    def group(group: Int): Option[String] = {
      Option(matcher.group(group)).filter(s => s != null && s.nonEmpty)
    }

    val topicFilter = IncludeList(group(1).getOrElse(".*"))
    val partitionFilter = group(2).map(_.toInt) match {
      case Some(partition) =>
        (p: Int) => p == partition
      case None =>
        val lowerRange = group(3).map(_.toInt).getOrElse(0)
        val upperRange = group(4).map(_.toInt).getOrElse(Int.MaxValue)
        (p: Int) => p >= lowerRange && p < upperRange
    }
```

Basically, if no single partition is provided, we use a range with `0` and `Int.MaxValue` as defaults.

Reply:
Thank you, changed implementation to this